[38기 김택수] ADD : category 별 전체 상품조회기능 추가 및 엔드포인트 통합

robinkim93 opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- category 별 전체 상품조회 기능 추가

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. main category 및 sub category 클릭 시 해당 category 내 product들을 전부 조회하는 기능
- path parameter로 구분해서 categoryid를 받아 query문에서 where절로 처리하였음

2. 불러온 데이터 낮은가격 및 높은가격 순으로 정렬하는 기능 추가
- Query parameter 이용해 낮은가격 및 높은가격순으로 데이터 정렬하는 기능 추가
- key값 별로 1은 랜덤, 2은 낮은가격순, 3은 높은가격순으로 정렬가능

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)

<br />

## :: 기타 질문 및 특이 사항
- 현재 모든 백엔드팀이 고민하고 있는 부분입니다. Query Parameter는 필터링에 따라서 들어오는 데이터가 달라지는데 달라지는 데이터를 가지고 어떻게 필터를 걸어야할지 고민입니다. (예를 들면, size만 필터를 걸었다면 Query Parameter로는 size와 size에 대한 값만 들어올거고, size와 color를 같이 필터를 걸었다면 두개의 값이 들어올텐데 이것에 대한 처리가 어려움이 있습니다.)

- Query Parameter 이용 시 if문을 통해 값이 어떤것이냐에 따라 쿼리문을 작성하여 변수에 담아 Dao단으로 전달했는데 이러한 방식이 올바른 방식인지 궁금합니다.

robinkim93 committed:
[ADD : category 별 전체 상품조회기능 추가]

robinkim93 committed:
[ADD : product 필터링 기능 구현]

lordmyshepherd commented on controller/categoriesController.js:
```diff
@@ -0,0 +1,32 @@
+const categoriesService = require("../service/categoriesService");
+
+const getMainCategoriesAllProducts = async (req, res, next) => {
+  const { maincategoriesId } = req.params;
+  const { sorttype } = req.query;
+  const getMainCategoriesAllProducts =
+    await categoriesService.getMainCategoriesAllProducts(
+      maincategoriesId,
+      sorttype
+    );
+
+  return res
+    .status(200)
+    .json({ getMainCategoriesAllProducts: getMainCategoriesAllProducts });
+};
+
+const getSubCategoriesAllProducts = async (req, res, next) => {
+  const { subcategoriesId } = req.params;
+  const { sorttype } = req.query;
+
+  const getSubCategoriesAllProducts =
+    await categoriesService.getSubCategoriesAllProducts(
+      subcategoriesId,
+      sorttype
+    );
+
+  return res
+    .status(200)
+    .json({ getSubCategoriesAllProducts: getSubCategoriesAllProducts });
+};
+
```
`getMainCategoriesAllProducts`와 `getSubCategoriesAllProducts` 두 가지 엔드포인트 모두 결국에는 상품 목록을 반환하는 것 같습니다.
그렇다면, 하나로 합치는 방법은 없을지 고민해보시면 좋을 것 같아요!

robinkim93 committed:
[MODIFY : 엔드포인트 통합]
- 기존 maincategory와 subcategory로 나눠져 있던 엔드포인트 하나로 통합
- 통합하는 과정에서 불필요한 category router 삭제
- Query Parameter로 받아온 뒤 if문으로 분기처리해서 main와 sub를 구분

lordmyshepherd reviewed (approved):
LGTM 👍🏻

lordmyshepherd commented on model/productsDao.js:
```diff
@@ -24,12 +24,28 @@ const categoriesProducts = async (categoriesId) => {
   `);
 };
 
+const getCategoriesAllProducts = async (categoriesId, sort) => {
+  return await myDataSource.query(`
+    SELECT 
+      p.name AS productName,
+      p.thumnail_image_url AS thumbnailImageUrl,
+      p.short_description AS shortDescription,
+      p.price AS price
+    FROM products AS p
+    INNER JOIN sub_categories AS sc ON p.sub_category_id = sc.id
+    INNER JOIN categories c ON sc.category_id = c.id
+    ${sort}
+    ${categoriesId}
+    `);
+};
```
아래와 같이 네이밍을 코드가 더 읽기 쉬워질 것 같습니다-!
```suggestion
const getProducts = async (whereClause, sort) => {
  return await myDataSource.query(`
    SELECT 
      p.name AS productName,
      p.thumnail_image_url AS thumbnailImageUrl,
      p.short_description AS shortDescription,
      p.price AS price
    FROM products AS p
    INNER JOIN sub_categories AS sc ON p.sub_category_id = sc.id
    INNER JOIN categories c ON sc.category_id = c.id
    ${sort}
    ${whereClause}
    `);
};
```

lordmyshepherd commented on service/productsService.js:
```diff
@@ -8,4 +8,31 @@ const getMdProducts = async (categoriesId) => {
 };
 
-module.exports = { getProducts, getMdProducts };
+const getCategoriesAllProducts = async (
```
함수명이 너무 긴 것 같습니다.
`getProducts` 이름 만으로도 의미 전달이 충분할 것 같습니다.

robinkim93 committed:
[MODIFY : Function naming 변경 및 Function parameter naming 수정]
- 기존 getCategoriesAllProducts 함수 네이밍이 길기 때문에 getAllCategories 로 변경 (충분히 의미 전달이 가능)
- Dao단에 넘겨주던 categoriesId 인자의 네이밍을 whereClause로 변경하여 DB에 가까운 인자로 바꿔줬음

lordmyshepherd reviewed (approved):
LGTM 👍🏻
